flink/src/main/java/org/apache/iceberg/flink/writer/PartitionWriter.java

```diff
@@ -0,0 +1,146 @@
+  private final Function<T, PartitionKey> keyGetter;
+  private final Map<PartitionKey, WrappedFileAppender<T>> writers;
+  private final long targetFileSize;
+  private final FileFormat fileFormat;
+  private final List<DataFile> completeDataFiles;
+
+  PartitionWriter(PartitionSpec spec,
+                  FileAppenderFactory<T> factory,
+                  Function<PartitionKey, EncryptedOutputFile> outputFileGetter,
+                  Function<T, PartitionKey> keyGetter,
+                  long targetFileSize,
+                  FileFormat fileFormat) {
+    this.spec = spec;
+    this.factory = factory;
+    this.outputFileGetter = outputFileGetter;
+    this.keyGetter = keyGetter;
+    this.writers = Maps.newHashMap();
+    this.targetFileSize = targetFileSize;
+    this.fileFormat = fileFormat;
+    this.completeDataFiles = Lists.newArrayList();
+  }
+
+  @Override
+  public void append(T record) throws IOException {
+    PartitionKey partitionKey = keyGetter.apply(record);
+    Preconditions.checkArgument(partitionKey != null, "Partition key shouldn't be null");
+
+    WrappedFileAppender<T> writer = writers.get(partitionKey);
+    if (writer == null) {
+      writer = createWrappedFileAppender(partitionKey);
+      writers.put(partitionKey, writer);
+    }
+    writer.fileAppender.add(record);
+
+    // Roll the writer if reach the target file size.
+    writer.currentRows++;
+    if (writer.currentRows % ROW_DIVISOR == 0 && writer.fileAppender.length() >= targetFileSize) {
+      closeCurrentWriter(writer);
+      writers.remove(partitionKey);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    for (WrappedFileAppender<T> wrap : writers.values()) {
+      closeCurrentWriter(wrap);
+      LOG.debug("Close file appender: {}, completeDataFiles: {}",
+          wrap.encryptedOutputFile.encryptingOutputFile().location(),
+          completeDataFiles.size());
+    }
+    this.writers.clear();
+  }
+
+  @Override
+  public List<DataFile> getCompleteFiles() {
+    if (completeDataFiles.size() > 0) {
+      return ImmutableList.copyOf(this.completeDataFiles);
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  @Override
+  public void reset() {
+    this.completeDataFiles.clear();
+  }
+
+  private WrappedFileAppender<T> createWrappedFileAppender(PartitionKey partitionKey) {
+    EncryptedOutputFile outputFile = outputFileGetter.apply(partitionKey);
+    FileAppender<T> appender = factory.newAppender(outputFile.encryptingOutputFile(), fileFormat);
+    return new WrappedFileAppender<>(partitionKey, outputFile, appender);
+  }
+
+  private void closeCurrentWriter(WrappedFileAppender<T> wrap) throws IOException {
+    DataFile dataFile = closeFileAppender(wrap.fileAppender, wrap.encryptedOutputFile, spec, wrap.partitionKey);
+    completeDataFiles.add(dataFile);
+  }
+
+  private static class WrappedFileAppender<T> {
+    private final PartitionKey partitionKey;
```

Comment:
We don't need store `partitionKey`, it is already in map?

Reply:
It's will be simpler to write the following logic with attached the `partitionKey` in WrappedFileAppender.